src/hotspot/share/gc/g1/g1MMUTracker.cpp

```diff
@@ -111,46 +111,55 @@ void G1MMUTracker::add_pause(double start, double end) {
   }
 }
 
+//                                          current_timestamp
+//                   GC events             /  pause_time
+//                 /     |     \          | /  /
+// -------------[----]-[---]--[--]--------|[--]-----> Time
+//              |         |                   |
+//              |         |                   |
+//              |<- limit |                   |
+//              |         |<- deficit         |
+//              |         ^                   |
+//              |                             |
+//              |<------- _time_slice   ----->|
+//
+// Focusing on GC events that are inside [limit, current_timestamp], we iterate
+// over them from the newest to the oldest (right-to-left in the diagram) and
+// try to locate the timestamp annotated with ^, so that the accumulated GC
+// time inside [deficit, current_timestamp], is equal to gc_budget. Next,
+// return `deficit - limit`.
+//
+// When there are no enough GC events, i.e. gc_budget has a surplus, return 0.
 double G1MMUTracker::when_sec(double current_timestamp, double pause_time) {
-  // If the pause is over the maximum, just assume that it's the maximum.
-  double adjusted_pause_time =
-    (pause_time > max_gc_time()) ? max_gc_time() : pause_time;
+  assert(pause_time > 0.0, "precondition");
 
-  // Earliest end time of a hypothetical pause starting now, taking pause_time.
-  double earliest_end_time = current_timestamp + adjusted_pause_time;
-  double gc_time_in_recent_time_slice = calculate_gc_time(earliest_end_time) + adjusted_pause_time;
+  // Clamp it by max
```

Comment:
```suggestion
    // If the pause is over the maximum, just assume that it's the maximum.
```
I would prefer the original comment as it explicitly states that this is a conscious decision.